src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1453,42 +1453,102 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
-                        warnings.Add(ErrorCode.WRN_InvalidMainSig, candidate.Locations.First(), candidate);
-                        continue;
+                        intOrVoidEntryPoints.Add(candidate);
+                        perCandidateBag.Free();
+                    }
+                    else if (IsTaskLike)
+                    {
+                        taskEntryPoints.Add((IsCandidate, candidate, perCandidateBag));
+                    }
+                    else
+                    {
+                        noMainFoundDiagnostics.Add(ErrorCode.WRN_InvalidMainSig, candidate.Locations.First(), candidate);
+                        noMainFoundDiagnostics.AddRange(perCandidateBag);
+                        perCandidateBag.Free();
                     }
+                }
+
+                var viableEntryPoints = ArrayBuilder<MethodSymbol>.GetInstance();
 
+                foreach (var candidate in intOrVoidEntryPoints)
+                {
                     if (candidate.IsGenericMethod || candidate.ContainingType.IsGenericType)
                     {
                         // a single error for partial methods:
-                        warnings.Add(ErrorCode.WRN_MainCantBeGeneric, candidate.Locations.First(), candidate);
+                        noMainFoundDiagnostics.Add(ErrorCode.WRN_MainCantBeGeneric, candidate.Locations.First(), candidate);
                         continue;
                     }
-
-                    // PROTOTYPE(async-main): CheckFeatureAvailability should be called on non-async methods 
-                    // that return Task or Task<T>
                     if (candidate.IsAsync)
                     {
-                        // PROTOTYPE(async-main): Get the diagnostic to point to a smaller syntax piece.
-                        // PROTOTYPE(async-main): Switch diagnostics around if the type is not Task or Task<T>
-                        CheckFeatureAvailability(candidate.GetNonNullSyntaxNode(), MessageID.IDS_FeatureAsyncMain, diagnostics);
+                        diagnostics.Add(ErrorCode.ERR_NonTaskMainCantBeAsync, candidate.Locations.First(), candidate);
+                        continue;
                     }
-
                     viableEntryPoints.Add(candidate);
                 }
 
-                if ((object)mainType == null || viableEntryPoints.Count == 0)
+                if (viableEntryPoints.Count == 0)
+                {
+                    foreach (var (IsValid, Candidate, SpecificDiagnostics) in taskEntryPoints)
+                    {
+                        if (Candidate.IsGenericMethod || Candidate.ContainingType.IsGenericType)
+                        {
+                            // a single error for partial methods:
+                            noMainFoundDiagnostics.Add(ErrorCode.WRN_MainCantBeGeneric, Candidate.Locations.First(), Candidate);
+                            continue;
+                        }
+
+                        if (!IsValid)
```

Comment:
Can these two `if` blocks be extracted to a helper method used in both the non-task-like and task-like candidates? (It would mean performing both checks for non-task-like candidates in the same loop, perhaps the first `foreach`, so the candidate is never added to `intOrVoidEntryPoints`.)